### PR TITLE
1172:Prevent scaling down deployments with multiple aliases

### PR DESCRIPTION
**Pull request by sreeramjayan**
This PR is to prevent a user from scaling down a deployment if it has multiple aliases.
Issue: https://github.com/zeit/now-cli/issues/1172

### CODE REVIEW

File: src/providers/sh/commands/scale.js [hunk 1]
```diff
@@ -244,6 +250,12 @@ async function run({ token, sh: { currentTeam } }) {
     return
   }
 
+  // Don't scale down the deployment if multiple aliases are linked to the same deploymentId
+  if (match.scale.current > scaleArg && duplicateDeployments.length > 1) {
+    console.log(`Scaling cannot be performed as there are multiple aliases linked to ${match.url}`)
+    return process.exit(0)
```
Comment (jamo) [inline src/providers/sh/commands/scale.js:256]: I wonder what the correct behavior should actually be:
probably we want to allow scaling up the new deployment, but not scale down the old one.

So with this approach, we would not even allow scaling up on alias if multiple aliases are pointing to the old URL.
So maybe this check should be just after scaling up the new instance, and then instead of scaling down the old one we would log this and proceed normally (or do we have a helper for that...)

**Comment by jamo** [comment]
Thanks for the PR, added one inline-comment, I think that we might want to approach this just a bit differently (see the comment).

**Comment by sreeramjayan** [comment]
@jamo 

> So with this approach, we would not even allow scaling up on alias if multiple aliases are pointing to the old URL.

This PR is to only prevent users from **scaling down** deployments that are tied to multiple aliases. I think https://github.com/zeit/now-cli/issues/1172 was raised to address issues that occur when a user tries to scale down a deployment tied to multiple aliases. 

I may have misunderstood the problem though.

**Comment by luisherranz** [comment]
That's right, my issue was about preventing the **scaling down** if deployment is tied to another alias with constrains.

This is the problematic scenario:

Aliases
- staging.domain.com -> project-123.now.sh (min: 3, max: 20, auto)
- production.domain.com -> project-123.now.sh (min: 3, max: 20, auto)

Create new deploy:
- project-456.now.sh

Alias it to staging to test it:
- staging.domain.com -> project-456.now.sh (min: 3, max: 20, auto)

This should scale up **project-456.now.sh** to at least 3, but should not scale down **project-123.now.sh** to 1 because it's still used for **production.domain.com**!

If testing went ok, use that deployment for production:
- production.domain.com -> project-456.now.sh (min: 3, max: 20, auto)

Now, we can scale down **project-123.now.sh** to 1, because it's not aliased anymore.

**Comment by luisherranz** [comment]
I'd even add a slightly more complex scenario, to make it as resilient as possible:

Aliases
- domain1.com -> project-123.now.sh (min: 3, max: 20, auto)
- domain2.com -> project-123.now.sh (min: 6, max: 10, auto)

**project-123.now.sh** should have internally this scaling settings right now: min: 6, max: 20, auto

Create new deploy:
- project-456.now.sh

Alias it to domain1:
- domain1.com -> project-456.now.sh

**project-456.now.sh** should have internally this scaling settings right now: min: 3, max: 20, auto
**project-123.now.sh** should have internally this scaling settings right now: min: 6, max: 10, auto

Alias it to domain2:
- domain2.com -> project-456.now.sh

**project-456.now.sh** should have internally this scaling settings right now: min: 6, max: 20, auto
**project-123.now.sh** should have internally this scaling settings right now: min: 0, max: 1, auto

So basically, deployments should have internal scaling settings where both minimum and maximum settings are the highest value among their current alias. Makes sense?

**Comment by sreeramjayan** [comment]
@luisherranz Thanks for the explanation. I guess I did misunderstand the issue. This PR prevents a person from scaling down an aliased deployment if another alias is tied to it.

For example:

- staging.domain.com -> project-123.now.sh (min: 3, max: 3)
- production.domain.com -> project-123.now.sh (min: 3, max: 3)

If I try to scale down an instance with
> now scale staging.domain.com 1

an error would be thrown since production.domain.com is also tied to project-123.now.sh

With the OSS plan I am on, it's hard to come up with a solution for your issue since auto scaling isn't provided.

**Comment by luisherranz** [comment]
I see! Thanks anyway @sreeramjayan :)

But maybe we could use your `duplicateDeployments` to add the logic I mentioned and calculate the new `min` and `max` before the `scale.setScale`.

@jamo could you please let me know where should I add tests for this? Maybe I can find time to do it myself in a couple of weeks.

**Comment by leo** [comment]
Closing in favor of https://github.com/zeit/now-cli/pull/1256 😊
